mempool/mempool.go

```diff
@@ -270,10 +304,19 @@ func (mem *Mempool) filterTxs(blockTxsMap map[string]struct{}) []types.Tx {
 	goodTxs := make([]types.Tx, 0, mem.txs.Len())
 	for e := mem.txs.Front(); e != nil; e = e.Next() {
 		memTx := e.Value.(*mempoolTx)
+		// Remove the tx if its alredy in a block.
 		if _, ok := blockTxsMap[string(memTx.tx)]; ok {
-			// Remove the tx since already in block.
+			// remove from clist
 			mem.txs.Remove(e)
 			e.DetachPrev()
+
+			// remove from mempool cache
+			// we only enforce "at-least once" semantics and
+			// leave it to the application to implement "only-once"
+			// via eg. sequence numbers, utxos, etc.
+			// NOTE: expects caller of filterTxs to hold the lock
+			// (so we can't use mem.removeTxFromCacheMap)
+			delete(mem.cacheMap, string(memTx.tx))
```

Comment:
IMO the replay protection is should either in tendermint (by ensuring that TXs can never be repeated) or TMSP side. Having both seems confusing, there should be a clear separation of responsibility.

Making tendermint do it seems prohibitive, there might be a use case for it out there and it will also be resource consuming (it will require a hash lookup of TX payload I imagine and ability to do this on all previous TXs?).. 

So I think the cleanest solution is just to let the TMSP app deal with this itself, by using something non-expensive as a nonce.